pilot/pkg/model/xds_cache.go

```diff
@@ -144,31 +164,90 @@ func newLru() simplelru.LRUCache {
 	return l
 }
 
-func (l *lruCache) Add(entry XdsCacheEntry, value *any.Any) {
+// assertUnchanged checks that a cache entry is not changed. This helps catch bad cache invalidation
+// We should never have a case where we overwrite an existing item with a new change. Instead, when
+// config sources change, Clear/ClearAll should be called. At this point, we may get multiple writes
+// because multiple writers may get cache misses concurrently, but they ought to generate identical
+// configuration. This also checks that our XDS config generation is deterministic, which is a very
+// important property.
+func assertUnchanged(existing *any.Any, replacement *any.Any) {
+	if features.EnableUnsafeAssertions {
+		if existing == nil {
+			// This is a new addition, not an update
+			return
+		}
+		if !cmp.Equal(existing, replacement, protocmp.Transform()) {
+			warning := fmt.Errorf("assertion failed, cache entry changed but not cleared: %v\n%v\n%v",
+				cmp.Diff(existing, replacement, protocmp.Transform()), existing, replacement)
+			panic(warning)
+		}
+	}
+}
+
+func (l *lruCache) Add(entry XdsCacheEntry, token CacheToken, value *any.Any) {
 	if !entry.Cacheable() {
 		return
 	}
 	l.mu.Lock()
 	defer l.mu.Unlock()
 	k := entry.Key()
-	l.store.Add(k, value)
+	cur, f := l.store.Get(k)
+	toWrite := cacheValue{value: value}
+	if f {
+		if token != cur.(cacheValue).token {
+			// entry may be stale, we need to drop it. This can happen when the cache is invalidated
+			// after we call Get.
+			return
+		}
+		// Otherwise, make sure we write the current token again. We don't change the key on writes; the
+		// same token will be used for a value until its invalidated
+		toWrite.token = cur.(cacheValue).token
```

Comment:
This allows concurrent writes to take the last value instead of stuck with the first written value. Generally, the guess is newer values > older values.

However, older != newer is bug, which we detect in asserts, so it *should* never matter